doc/dev/static_type_checking.md

```diff
@@ -0,0 +1,894 @@
+    - include the `py.typed` file in the
+      MANIFEST.in ([example](https://github.com/Azure/azure-sdk-for-python/blob/main/sdk/core/azure-core/MANIFEST.in))
+
+2) Add type hints anywhere in the source code where unit tests are worth writing. Consider typing/mypy as "free" tests
+   for your library so focusing typing on high density/important areas of the code helps in detecting bugs.
+
+### Types usable in annotations
+
+Almost anything can be used as a type in annotations.
+
+1) Basic types like `int`, `str`, `bytes`, `float`, `bool`
+2) Classes or abstract base classes - whether user-defined, from the standard library
+   like [collections](https://docs.python.org/3/library/collections.html)
+   or [collections.abc](https://docs.python.org/3/library/collections.abc.html), or external packages
+3) Types from the [typing](https://docs.python.org/3/library/typing.html)
+   or [typing_extensions](https://github.com/python/typing/tree/master/typing_extensions) modules
+4) Built-in generic types, like `list` or `dict`*.
+   > *Note: Support for generic types was not added until Python 3.9 (or 3.7+ with `from __future__ import annotations` import). This lets you pass in generic `list[str]` as a type hint rather than `typing.List[str]`.
+
+Here are a few considerations and notes on Python version support:
+
+- With [PEP585](https://peps.python.org/pep-0585/) and [PEP563](https://peps.python.org/pep-0563/), importing certain
+  generic collection types from `typing` has been [deprecated](https://peps.python.org/pep-0585/#implementation) in
+  favor for using the types in `collections.abc` or generic type hints (like `list`) from the standard library. At the
+  time of writing, the Python SDK supports 3.6+ and so we must still import these types from `typing` or check the
+  Python version at runtime before importing (Note that the `collections.abc` types did not have [] notation added until
+  3.9)
+
+```python
+import sys
+
+if sys.version_info < (3, 9):
+    from typing import Sequence
+else:
+    from collections.abc import Sequence
+```
+
+- The `typing-extensions` library backports types that are introduced only in later versions of Python (e.g. `Protocol`
+  which was introduced in Python 3.8) so that they can be used as early as Python 3.6.
+
+```python
+from typing_extensions import Protocol
+```
+
+Our azure-core library takes a dependency on `typing-extensions` so for most client libraries the dependency is already
+implicit. See the [typing-extensions](https://github.com/python/typing/tree/master/typing_extensions) docs to check what
+has been backported.
+
+## Install and run mypy on your client library code
+
+Our Python SDK repo has the version of mypy that we run in CI pinned to a specific version. This is to avoid surprise
+typing errors raised when a new version of `mypy` ships. All client libraries in the Python SDK repo are automatically
```

Comment:
What version should I isntall at the minimal? Should it be written?

Reply:
Included it. I don't think we'll bump the version often enough that it will make it hard to keep this doc up-to-date.

Reply:
Included it. I don't think we'll bump the version often enough that it will make it hard to keep this doc up-to-date.